File: 08/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const maxN = 700000

// Максимальное k, для которого существуют k-интересные числа <= maxN
// 2*3*4*5*6*7*8*9 = 362880 <= 700000
// 2*3*4*5*6*7*8*9*10 = 3628800 > 700000
// Значит maxK = 9
const maxK = 9

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 1<<20)
	writer := bufio.NewWriterSize(os.Stdout, 1<<20)
	defer writer.Flush()

	// Предвычисляем префиксные суммы k-интересных чисел
	prefixSums := precompute()

	// Читаем количество запросов
	line, _ := reader.ReadString('\n')
	q, _ := strconv.Atoi(strings.TrimSpace(line))

	// Обрабатываем запросы
	for i := 0; i < q; i++ {
		line, _ = reader.ReadString('\n')
		parts := strings.Fields(strings.TrimSpace(line))
		if len(parts) < 3 {
			continue
		}
		k, _ := strconv.Atoi(parts[0])
		l, _ := strconv.Atoi(parts[1])
		r, _ := strconv.Atoi(parts[2])

		result := query(prefixSums, k, l, r)
		writer.WriteString(fmt.Sprintf("%d\n", result))
	}
}

// precompute генерирует все k-интересные числа и строит префиксные суммы
func precompute() [][]int32 {
	// prefixSums[k][n] = количество k-интересных чисел от 1 до n
	prefixSums := make([][]int32, maxK+1)
	for k := 1; k <= maxK; k++ {
		prefixSums[k] = make([]int32, maxN+1)
	}

	// Генерируем k-интересные числа для k >= 2 и сразу помечаем
	generate(1, 2, 0, maxN, prefixSums)

	// Все числа >= 2 являются 1-интересными
	for n := 2; n <= maxN; n++ {
		prefixSums[1][n] = 1
	}

	// Преобразуем в префиксные суммы
	for k := 1; k <= maxK; k++ {
		for n := 1; n <= maxN; n++ {
			prefixSums[k][n] += prefixSums[k][n-1]
		}
	}

	return prefixSums
}

// generate рекурсивно генерирует все произведения возрастающих последовательностей множителей
func generate(product, minFactor, depth, limit int, prefixSums [][]int32) {
	for factor := minFactor; product*factor <= limit; factor++ {
		newProduct := product * factor
		newDepth := depth + 1

		if newDepth >= 2 && newDepth <= maxK {
			prefixSums[newDepth][newProduct] = 1
		}

		generate(newProduct, factor+1, newDepth, limit, prefixSums)
	}
}

// query возвращает количество k-интересных чисел в диапазоне [l, r]
func query(prefixSums [][]int32, k, l, r int) int32 {
	if k > maxK || l > maxN {
		return 0
	}
	if r > maxN {
		r = maxN
	}
	if l < 1 {
		l = 1
	}
	return prefixSums[k][r] - prefixSums[k][l-1]
}

File: 08/main.rs
use std::io::{self, BufRead, BufWriter, Write};

const MAX_N: usize = 700000;

// Максимальное k, для которого существуют k-интересные числа <= maxN
// 2*3*4*5*6*7*8*9 = 362880 <= 700000
// 2*3*4*5*6*7*8*9*10 = 3628800 > 700000
// Значит maxK = 9
const MAX_K: usize = 9;

fn main() {
    let stdin = io::stdin();
    let mut reader = stdin.lock();
    let stdout = io::stdout();
    let mut writer = BufWriter::new(stdout.lock());

    // Предвычисляем префиксные суммы k-интересных чисел
    let prefix_sums = precompute();

    // Читаем количество запросов
    let mut line = String::new();
    reader.read_line(&mut line).unwrap();
    let q: usize = line.trim().parse().unwrap();

    // Обрабатываем запросы
    for _i in 0..q {
        line.clear();
        reader.read_line(&mut line).unwrap();
        let parts: Vec<&str> = line.trim().split_whitespace().collect();
        if parts.len() < 3 {
            continue;
        }
        let k: usize = parts[0].parse().unwrap();
        let l: usize = parts[1].parse().unwrap();
        let r: usize = parts[2].parse().unwrap();

        let result = query(&prefix_sums, k, l, r);
        writeln!(writer, "{}", result).unwrap();
    }
    writer.flush().unwrap();
}

// precompute генерирует все k-интересные числа и строит префиксные суммы
fn precompute() -> Vec<Vec<i32>> {
    // prefix_sums[k][n] = количество k-интересных чисел от 1 до n
    let mut prefix_sums = vec![vec![0i32; MAX_N + 1]; MAX_K + 1];

    // Генерируем k-интересные числа для k >= 2 и сразу помечаем
    generate(1, 2, 0, MAX_N, &mut prefix_sums);

    // Все числа >= 2 являются 1-интересными
    for n in 2..=MAX_N {
        prefix_sums[1][n] = 1;
    }

    // Преобразуем в префиксные суммы
    for k in 1..=MAX_K {
        for n in 1..=MAX_N {
            prefix_sums[k][n] += prefix_sums[k][n - 1];
        }
    }

    prefix_sums
}

// generate рекурсивно генерирует все произведения возрастающих последовательностей множителей
fn generate(
    product: usize,
    min_factor: usize,
    depth: usize,
    limit: usize,
    prefix_sums: &mut Vec<Vec<i32>>,
) {
    let mut factor = min_factor;
    while product * factor <= limit {
        let new_product = product * factor;
        let new_depth = depth + 1;

        if new_depth >= 2 && new_depth <= MAX_K {
            prefix_sums[new_depth][new_product] = 1;
        }

        generate(new_product, factor + 1, new_depth, limit, prefix_sums);
        factor += 1;
    }
}

// query возвращает количество k-интересных чисел в диапазоне [l, r]
fn query(prefix_sums: &[Vec<i32>], k: usize, l: usize, r: usize) -> i32 {
    if k > MAX_K || l > MAX_N {
        return 0;
    }
    let r = r.min(MAX_N);
    let l = l.max(1);
    if l > r {
        return 0;
    }
    prefix_sums[k][r] - prefix_sums[k][l - 1]
}








